### DISCUSSION TITLE
How to handle sync endpoints with multithreading most appropriately

### DISCUSSION THREAD

**Discussion by benedikt-mue** [fastapi/fastapi#13101] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import Depends, FastAPI

app = FastAPI()

def prepare_question_data(question_id):
    # pseudo code:
    # - get data from postgres db
    # - transform and process data with pandas

@app.get("/all_answers")
aysnc def all_answers():
    with ThreadPoolExecutor() as executor:
        results = list(
            executor.map(
                lambda question_id: prepare_question_data(
                    question_id=question_id,
                ),
                question_ids,
            )
        )

# Debugger
if __name__ == "__main__":
    import uvicorn
    uvicorn.run(
        "__main__:app",
        host="127.0.0.1",
        workers=1,
        port=8000,
        log_config=None,
    )
```

### Description

Pls help to settle discussion a coworker and I have about the correct setup for a fastApi endpoint that fetches some data from postgres and processes it using pandas. I did some load testing and it was obviously not great. 
Endpoint specs:
- question_ids is usually around 30-50 in size
- the datasets are fairly small 
- the individual queries take <1s 
- a single call to the endpoint finishes in about 4 sec
- there is not much load on this endpoint (<5 concurrent request max)
- use a sync endpoint instead of aysnc to not block event loop
- spawn multithreading for i/o bound db calls inside the sync endpoint 
- process the data in the threads (this is not giving any speedups, I know, and I think this is also bad for the Api performance itself)
Others specs: 
- max <50 concurrent requests on entire Api
- several, other async endpoints (that also do a little bit of data wrangling)

objectives:
- handle the number of concurrent requests
- speedup the endpoints (if possible)
- do not scale to multiple instances, if no necessary 

questions:
1.
- does fastapi/uvicorn´s event loop stays unblocked with increasing requests, if, e.g. max threads are already allocated
- or should I limit the No of threadpools for each request to not block fastApi to spawn additional thread for new requests?
2.
- would it make sense to use multiprocessing instead to avoid clogging 
- or does it make sense to rather increase the no of workers?

any feedback is much appreciated! 

### Operating System

Linux, Other

### Operating System Details

Running as dockerized Azure Webapp (machine size is not decided as of yet)

### FastAPI Version

0.115.0

### Pydantic Version

2.0.2

### Python Version

3.9

**Comment by YuriiMotov** [comment, marked as answer]
The way you are running blocking code inside an async function is wrong.
`list(executor.map(...))` will block async loop until all sync functions have been finished execution. So, your server will not be able to process other requests during this time.

To fix this, run sync code using this approach:
```
@app.get("/all_answers")
async def all_answers():
    loop = asyncio.get_running_loop()
    with ThreadPoolExecutor(max_workers=50) as executor:
        tasks = [
            loop.run_in_executor(executor, prepare_question_data, question_id)
            for question_id in question_ids
        ]
        results = await asyncio.gather(*tasks)
    return results
```

To futher optimize this code, create single `ThreadPoolExecutor` instance on you application run (in lifespan) and use it instead of creating new pool every time

**Reply by YuriiMotov** [reply]
To futher optimize this code, create single `ThreadPoolExecutor` instance on you application run (in lifespan) and use it instead of creating new pool every time.
```
import asyncio
from concurrent.futures import ThreadPoolExecutor
from contextlib import asynccontextmanager
import time
from typing import cast

from fastapi import FastAPI, Request


@asynccontextmanager
async def lifespan(_: FastAPI):
    with ThreadPoolExecutor(max_workers=50) as executor:
        yield {"executor": executor}

app = FastAPI(lifespan=lifespan)

def prepare_question_data(question_id):
    time.sleep(5)
    return f"{question_id} data"

@app.get("/all_answers")
async def all_answers(request: Request):
    question_ids = [1, 2, 3, 4, 5, 6, 7]

    executor = cast(ThreadPoolExecutor, request.state.executor)
    loop = asyncio.get_running_loop()
    tasks = [
        loop.run_in_executor(executor, prepare_question_data, question_id)
        for question_id in question_ids
    ]
    results = await asyncio.gather(*tasks)
    return results
```

**Reply by benedikt-mue** [reply]
thy for the reply. 
Regarding the first snippet, wouldn´t it, as I wrote, also just make sense to use a sync endpoint ?
Regarding the second one, is it okay to spawn up to 50 threads, wouldn´t that also cause issues?

**Reply by YuriiMotov** [reply]
Yes, making the endpoint function sync will help, but it's a bit less efficient.

And yes, you need to adjust the size of thread pool

**Reply by benedikt-mue** [reply]
One worker can handle 40 threads, so if I use two, I could go upto 80? Would it be okay to always scale to the max number? 
I cannot find any docs about this, so sorry for those questions. If you have any supporting docs, it would be much appriciated.

**Reply by YuriiMotov** [reply]
When you say "worker", do you mean Uvicorn worker? 
Workers are not limited by 40 threads. You can set `ThreadPoolExecutor(max_workers=500)` and it will likely to work well

**Reply by benedikt-mue** [reply]
Yea, I meant the uvicorn worker, it is set to 40 by default. I have read that it is set to a lower number as pyhton sucks in handling too many threads, so I am just curious if by increasing the number of workers, increasing the number of threads is also beneficial?
